src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverter.cs

```diff
@@ -55,8 +58,19 @@ internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert,
             }
 
             Debug.Assert(options.UnknownTypeHandling == JsonUnknownTypeHandling.JsonNode);
-            value = JsonNodeConverter.Instance.Read(ref reader, typeToConvert, options)!;
-            // TODO reference lookup for JsonNode deserialization.
+
+            JsonNode node = JsonNodeConverter.Instance.Read(ref reader, typeToConvert, options)!;
+
+            if (options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.Preserve &&
+                JsonSerializer.TryHandleReferenceFromJsonNode(ref reader, ref state, node, out referenceValue))
+            {
+                value = referenceValue;
+            }
+            else
+            {
+                value = node;
+            }
```

Comment:
could this be collapsed into something like: `JsonNodeConverter.Instance.ReadValueOrReference(....)` which would handle this for you?

Reply:
Extracting any `$ref` metadata from the deserialized value should not be the responsibility of the converter. In any case, this logic is part of the ObjectConverter, it's not part of the deserialization hot path.

Reply:
it still could be `WhateverClass.ReadValueOrReference(converter, ...)`